### PR TITLE
dnsdist: Refactoring of the TCP stack

### LINKED ISSUE #4814: dnsdist TCP stack needs improving
- Program: dnsdist 
 - Issue type: Bug report

### Description
The current dnsdist TCP stack has some weaknesses particularly when it is used in front of recursors. Due to the way dnsdist handles distributing queries over its tcp threads it can cause a "jam" of (very) slow queries on a certain thread. When it then assigns a "normal" query to this thread this query might timeout. Another case is that the queue for a certain thread can fill up the entire global TCP queue while the other thread still have plenty of time for processing queries causing queries to get dropped since the queue is "full".

These effects can be mitigated by spawning *alot* of tcp threads or setting the tcp recv timeouts on the server very low, but neither solution is very desirable.

### Environment
 - Operating system: Any 
 - Software version: 1.0.0, 1.1.0-beta2, git (probably)

### Steps to reproduce
The easiest way to reproduce is to fire a bunch of known slow queries at dnsdist. dnsdist will then stop responding properly to normal queries as well or at least encountering failures at random (although I managed to temporarily break it completely with about 1 query every 2 seconds in my case).

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

This PR completely changes the way dnsdist handles TCP connections, moving to an event-based mode that allows a single thread to handle a large number of connections, instead of only one. It makes DNS over TCP and DNS over TLS much more scalable.
This needs a very serious review, lots of testing, including weird corner-cases like TCP Fast Open, etc.. 
The documentation needs to be updated, and probably the way we handle the number of TCP threads to start.

Based on #7526. Hopefully fixes #4814.

IMHO merging this PR calls for the move to the 1.4.x branch at least, perhaps 2.0.x.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by rgacogne** [comment]
Rebased to fix a conflict.

File: pdns/dnsdist-tcp.cc [hunk 1]
```diff
@@ -201,544 +260,863 @@ void TCPClientCollection::addTCPClientThread()
...
+      return;
+    }
+    throw std::runtime_error("Error while reading from the TCP acceptor pipe (" + std::to_string(pipefd) + ") in " + std::string(isNonBlocking(pipefd) ? "non-blocking" : "blocking") + " mde:" + strerror(errno));
+  }
+  else if (got != sizeof(citmp)) {
+    throw std::runtime_error("Partial read while reading from the TCP acceptor pipe (" + std::to_string(pipefd) + ") in " + std::string(isNonBlocking(pipefd) ? "non-blocking" : "blocking") + " mode");
+  }
+
+  g_tcpclientthreads->decrementQueuedCount();
+  auto ci = std::move(*citmp);
```
Comment (chbruyand) [inline pdns/dnsdist-tcp.cc]: I was wondering if we could get rid of `ci` and just construct `state` directly by moving `citmp`'s content

File: pdns/dnsdist-tcp.cc [hunk 2]
```diff
@@ -201,544 +260,863 @@ void TCPClientCollection::addTCPClientThread()
...
+
+  setThreadName("dnsdist/tcpClie");
+
+  TCPClientThreadData data;
+
+  data.mplexer->addReadFD(pipefd, handleIncomingTCPQuery, &data);
+  time_t lastTCPCleanup = time(nullptr);
+  time_t lastTimeoutScan = time(nullptr);
+  struct timeval now;
+  gettimeofday(&now, 0);
```
Comment (chbruyand) [inline pdns/dnsdist-tcp.cc]: Maybe we can just init `lastTCPCleanup` and `lastTimeoutScan` with `now.tv_sec`

File: pdns/tcpiohandler.hh [hunk 3]
```diff
@@ -176,6 +193,77 @@ public:
...
     }
   }
+
+  /* Tries to read exactly toRead bytes into the buffer, starting at position pos.
+     Updates pos everytime a successful read occurs,
+     throws an std::runtime_error in case of IO error,
+     return Done when toRead bytes have been read, needRead or needWrite if the IO operation
+     would block.
+  */
+  IOState tryRead(std::vector<uint8_t>& buffer, size_t& pos, size_t toRead)
```
Comment (chbruyand) [inline pdns/tcpiohandler.hh]: Wow, looks like the same code as in `dnsdist-tcp.cc`

File: pdns/dnsdist-tcp.cc [hunk 4]
```diff
@@ -201,544 +260,863 @@ void TCPClientCollection::addTCPClientThread()
...
+}
+
+static void handleNewIOState(std::shared_ptr<IncomingTCPConnectionState>& state, IOState iostate, const int fd, FDMultiplexer::callbackfunc_t callback, boost::optional<struct timeval> ttd)
+{
+  //cerr<<"in "<<__func__<<" for fd "<<fd<<", last state was "<<(int)state->d_lastIOState<<", new state is "<<(int)iostate<<endl;
+
+  if (state->d_lastIOState == IOState::NeedRead && iostate != IOState::NeedRead) {
+    state->d_threadData.mplexer->removeReadFD(fd);
+    //cerr<<__func__<<": remove read FD "<<fd<<endl;
+    state->d_lastIOState = IOState::Done;
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsdist-tcp.cc]: But if the next state (`iostate`) is `NeedWrite` we need to ask the multiplexer to let us know when the FD is writable, right?

File: pdns/tcpiohandler.hh (hunk 3 above)
Comment (rgacogne) [reply pdns/tcpiohandler.hh]: Yes :-) That's why the one in `dnsdist-tcp.cc` has this comment:

> // XXX could probably be implemented as a TCPIOHandler

but I haven't looked into it yet. It would be nice but it can wait a bit.

File: pdns/dnsdist-tcp.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (chbruyand) [reply pdns/dnsdist-tcp.cc]: Indeed, got it.

**Comment by rgacogne** [comment]
Just pushed a few commits, the only important one being 705cf3caf8544fe681567fc5501114a3ace7385d that fixes a nice bug with DoT, uncovered while testing with https://github.com/DNS-OARC/flamethrower

**Comment by rgacogne** [comment]
Rebased to fix conflicts.
